Update localization document with image tool (#1833)
google/automotive-design-compose: docs/_docs/localization/index.md

# Localization

We are developing tools for developers to localize the figma design into different languages more easily.

The first introduced tool is a figma plugin that generates a strings xml file containing the strings
used by the figma design.

> The second tool is a figma plugin that literates all the images, let user assign a proper name and

Reviewer: Is "literates" supposed to be "iterates"?

Author: Yes. 😅

> download selected images as a `drawable.zip` file.

Reviewer: Suggestion: "The second tool is a figma plugin that iterates through all the images, then lets a user assign a proper name and download selected images as a `drawable.zip` file."

## 1. Run the localization plugin

> Both tools are now published public as part of the figma plugin [Automotive Design For Compose][1].

Reviewer: ```suggestion
Both tools are now published publicly as part of the figma plugin [Automotive Design For Compose][1].
```

   ```

3. Install the figma desktop app. Figma will need to read the plugin code saved as a local file. You
   can download it from the [Figma downloads page][3].

4. Import the plugin from manifest and you are ready to run the plugin. The file path is:

   ```text
   automotive-design-compose/support-figma/extended-layout-plugin/manifest.json
   ```

   <img src="./ImportPluginFromManifest.png">

5. Run the plugin under development through:
   **Figma Main Menu -> Plugins -> Development -> Automotive Design For Compose -> Localization Tool**

## 2. String resource tool

### 2.1 Generate string resources

#### 2.1.1 Strings generation options

Currently the plugin provides three options to exclude text nodes from translations.

> <img src="./LocalizationOptions.png">

Reviewer: I remember an issue with the public site when these images were not in the proper folder, do you know if that's still the case?

Author: I moved the md file under the folder. That is why I don't need this any more.

The plugin also provides a **Clear String Resource** tool to clear the string resource names, char
limits and descriptions from the plugin data.

#### 2.3.2 It supports mixed styles

For a text node with a single style, the plugin generates a string resource. And for a text node with
mixed styles, the plugin generates a string array resource with each item mapping to the text segments
with different styles.

## 3. Image resource tool

Images play an important role in the figma design. Figma supports PNG, JPG and GIF files up to 4k
in width and height as background fills or strokes. In some cases, images need localization too,
for example, the stop sign looks different in different countries.

As a supplemental tool to the string resource tool, it works in a similar way that the plugin goes
through all the images, allows user to assign a proper resource name to the image which in the
Android resource framework is the image file name and it writes the resource name to the figma

> plugin data which the DesignCompose uses later to load the image resource.

Reviewer: Suggestion: "As a supplemental tool to the string resource tool, the image resource tool works in a similar way. It goes through all the images and allows a user to assign a proper resource name to the image to match the image file name in the Android resource framework. It then writes the resource name to the Figma plugin data, which DesignCompose uses later to load the image resource."

### 3.1 Download images

Run `Export Images` under the `Localization Tool`. The plugin will display a list of images used by

> the figma design. Each image will have `Image hash`, `Image res name`, the image and a list of

Reviewer: ```suggestion
the Figma design. Each image will have `Image hash`, `Image res name`, the image and a list of
```

`node id`s which are clickable to show the node with the image.

<img src="./ImageTool.png">

Each image has a checkbox which is unchecked by default. When checked, the image resource name will
be written to the shared plugin data with a specific namespace fetched by DesignCompose and the

> image will be exported when download.

Reviewer: ```suggestion
image will be exported when downloaded.
```

The image res name is an edit box which allows you to rename the auto created image resource name
to a more meaningful name. Note: you need to check the destination drawable folder you're planning
to place the images manually and avoid using the same resource name if the images are different.

After reviewing, renaming and picking the images for localization, you can click the `Download`
button to download the selected images as the `drawable.zip` file.

<img src="./SaveImage.png">

> Note: Although this tool can export GIF images, DesignCompose doesn't support GIF now. Figma REST api

Reviewer: ```suggestion
Note: Although this tool can export GIF images, DesignCompose currently doesn't support GIF. Figma's REST API
```

returns both `imageRef` and `gifRef` but DesignCompose only renders the `imageRef` which is a static

> image(see `Paint` properties <https://www.figma.com/developers/api#files-types>). The plugin api

Reviewer: ```suggestion
image(see `Paint` properties <https://www.figma.com/developers/api#files-types>). The plugin API
```

can only get the image hash(i.e. the `gifRef`), so localization doesn't work for GIFs at this point.

### 3.2 Integrate with the app and test

To integrate with the app, unzip the `drawable.zip` file and place all the images under the app's

> `res/drawable` folder. For each of image, place localized images under `res/drawable-[LOCALE]` folder.

Reviewer: ```suggestion
`res/drawable` folder. For each of the images, place localized images under `res/drawable-[LOCALE]` folder.
```

In the following example, we will add localized stop sign for `zh` and `fr`.

<img src="./LocalizedImage.png">

Update the figma dcf file, rebuild the app and deploy.

When the device's locale changes to Simplified Chinese, the app should use the localized image for
the stop sign from `drawable-zh`.
<img src="./ImageLocalizedResult.png">

Also, similar to 2.2.3, we can re-enable the live update for images by unchecking the
`Use local resources if exists` option in design switcher.

### 3.3 What is not supported

> Android recommend to create different versions of images for different screen densities. This tool

Reviewer: ```suggestion
Android recommends creating different versions of images for different screen densities. This tool
```

> only download the images in their original sizes and we suggest acquiring the other scales of

Reviewer: ```suggestion
only downloads the images in their original sizes and we suggest acquiring the other scales of
```
